src/wp-includes/l10n.php

```diff
@@ -2006,6 +2006,107 @@ function wp_get_word_count_type() {
+ *     @type array  $shortcodes                         Optional. Array of shortcodes that should be removed
+ *                                                      from the text.
+ * }
+ * @return int The word or character count.
+ */
+function wp_word_count( $text, $type, $settings = array() ) {
+	$defaults = array(
+		'html_regexp'                        => '/<\/?[a-z][^>]*?>/i',
+		'html_comment_regexp'                => '/<!--[\s\S]*?-->/',
+		'space_regexp'                       => '/&nbsp;|&#160;/i',
+		'html_entity_regexp'                 => '/&\S+?;/',
+		'connector_regexp'                   => "/--|\x{2014}/u",
+		'remove_regexp'                      => "/[\x{0021}-\x{0040}\x{005B}-\x{0060}\x{007B}-\x{007E}\x{0080}-\x{00BF}\x{00D7}\x{00F7}\x{2000}-\x{2BFF}\x{2E00}-\x{2E7F}]/u",
+		'astral_regexp'                      => "/[\x{010000}-\x{10FFFF}]/u",
+		'words_regexp'                       => '/\S\s+/u',
+		'characters_excluding_spaces_regexp' => '/\S/u',
+		'characters_including_spaces_regexp' => "/[^\f\n\r\t\v\x{00AD}\x{2028}\x{2029}]/u",
+		'shortcodes'                         => array(),
+	);
+
+	$count = 0;
+
+	if ( '' === trim( $text ) ) {
+		return $count;
+	}
+
+	$settings = wp_parse_args( $settings, $defaults );
+
+	// If there are any shortcodes, add this as a shortcode regular expression.
+	if ( is_array( $settings['shortcodes'] ) && ! empty( $settings['shortcodes'] ) ) {
+		$settings['shortcodes_regexp'] = '/\\[\\/?(?:' . implode( '|', $settings['shortcodes'] ) . ')[^\\]]*?\\]/';
+	}
+
+	// Sanitize type to one of three possibilities: 'words', 'characters_excluding_spaces' or 'characters_including_spaces'.
+	if ( 'characters_excluding_spaces' !== $type && 'characters_including_spaces' !== $type ) {
+		$type = 'words';
+	}
+
+	$text .= "\n";
+
+	// Replace all HTML with a new-line.
+	$text = preg_replace( $settings['html_regexp'], "\n", $text );
+
+	// Remove all HTML comments.
+	$text = preg_replace( $settings['html_comment_regexp'], '', $text );
+
+	// If a shortcode regular expression has been provided use it to remove shortcodes.
+	if ( ! empty( $settings['shortcodes_regexp'] ) ) {
+		$text = preg_replace( $settings['shortcodes_regexp'], "\n", $text );
+	}
+
+	// Normalize non-breaking space to a normal space.
+	$text = preg_replace( $settings['space_regexp'], ' ', $text );
+
+	if ( 'words' === $type ) {
+		// Remove HTML Entities.
+		$text = preg_replace( $settings['html_entity_regexp'], '', $text );
+
+		// Convert connectors to spaces to count attached text as words.
+		$text = preg_replace( $settings['connector_regexp'], ' ', $text );
+
+		// Remove unwanted characters.
+		$text = preg_replace( $settings['remove_regexp'], '', $text );
+	} else {
+		// Convert HTML Entities to "a".
+		$text = preg_replace( $settings['html_entity_regexp'], 'a', $text );
+
+		// Remove surrogate points.
+		$text = preg_replace( $settings['astral_regexp'], 'a', $text );
```

Comment:
something doesn’t match between the comment and the PCRE pattern.

 - all surrogate halves are in the basic multilingual plane.
 - surrogates are illegal in UTF-8.
 - the astral plane includes an incredible variety of content.

it seems like the goal is to replace content we don’t recognize with an `a`, hoping that strings of `a` letters will be counted as a single word.

if we want to consider surrogates though, we should treat them as invalid UTF-8 and ignore them, or call `wp_scrub_utf8()`/`mb_scrub()` beforehand to eliminate invalid text.

one note: although UTF-16 encodes all characters from the astral plane using surrogate pairs, these are not to be expected here where UTF-8 is the norm.